integration/store_gateway_limits_hit_test.go

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+//go:build requires_docker
+// +build requires_docker
+
+package integration
+
+import (
+	"context"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/gogo/status"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/dskit/grpcclient"
+	"github.com/grafana/e2e"
+	e2ecache "github.com/grafana/e2e/cache"
+	e2edb "github.com/grafana/e2e/db"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
+	grpc_metadata "google.golang.org/grpc/metadata"
+
+	"github.com/grafana/mimir/integration/e2emimir"
+	"github.com/grafana/mimir/pkg/storegateway"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+const (
+	blockRangePeriod = 5 * time.Second
+	ingesterTag      = "ingester"
+	distributorTag   = "distributor"
+	storeGatewayTag  = "store-gateway"
+	querierTag       = "querier"
+	orgID            = "test"
+	series1          = "series_1"
+	series2          = "series_2"
+	series3          = "series_3"
+	series4          = "series_4"
+)
+
+func TestStoreGateway_StoreGatewayLimitHit(t *testing.T) {
+	scenario, mimirServices, client := createClient(t, map[string]string{"-querier.max-fetched-series-per-query": "3"})
+	defer scenario.Close()
+
+	timeStamps, timeSeries := createTimeSeries(t, []string{series1, series2, series3, series4})
+
+	pushTimeSeries(t, client, timeSeries[series1])
+	pushTimeSeries(t, client, timeSeries[series2])
+	pushTimeSeries(t, client, timeSeries[series3])
+	pushTimeSeries(t, client, timeSeries[series4])
+
+	// ensures that data will be queried from store-gateway, and not from ingester: at this stage 1 block of data should be stored
+	waitUntilShippedToStorage(t, mimirServices, 1)
+
+	cfg := grpcclient.Config{}
+	flagext.DefaultValues(&cfg)
+	reg := prometheus.NewPedanticRegistry()
+	factory := storegateway.NewStoreGatewayClientFactory(cfg, reg)
+	storeGatewayClient, err := factory(mimirServices[storeGatewayTag].GRPCEndpoint())
+	require.NoError(t, err)
+
+	ctx := prepareGRPCContext()
+	req := &storepb.SeriesRequest{
+		MinTime: timeStamps[series1].Unix() * 1000,
+		MaxTime: timeStamps[series4].Add(1*time.Hour).Unix() * 1000,
+		Matchers: []storepb.LabelMatcher{
+			{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_.+"},
+		},
+	}
+	stream, err := storeGatewayClient.(*storegateway.ClientImpl).Series(ctx, req)
+	require.NoError(t, err)
+	require.NotNil(t, stream)
```

Comment:
Integration tests should focus on end-to-end behaviour. We shouldn't call store-gateway via gRPC, but run a normal query and assert on the actual result / error we get.

In this test we want to focus on the limit hit in the store-gateway. I think the solution is easy: just set the limit `-querier.max-fetched-series-per-query` to the store-gateway and **not** the querier. Since it's passed as a CLI flag, you can pass it to the store-gateway but not the querier.

In the test testing this limit in the querier, you can pass it to the querier but not the store-gateway.

The two tests ("hit limit in the querier" and "hit limit in the store-gateway") can also be written in the same `TestXXX()` function, because they should be the exact same, but passing the limit to one component or the other (so a normal table test should be fine).